src/app/data-planes/sources.ts

```diff
@@ -155,5 +156,31 @@ export const sources = (api: KumaApi, can: Can) => {
         size,
       }), can('use zones'))
     },
+
+    '/dataplanes/with-status-polling': async (params) => {
```

Comment:
You don't want `async` here

Reply:
Interesting. Why does this change the behavior of DataSource?

---

Unrelated to that issue, I think there is another reason we can’t use the retry mechanism for this as it’s currently built.

In this view, we want to poll `/dataplanes` until we no longer find any offline Dataplanes of the 10 that are fetched. _While_ this takes place, however, we want to show the current Dataplanes. But with the retry mechanism being exception-based, we can’t actually return the data _and_ trigger a retry. I’ve changed the approach to polling in https://github.com/kumahq/kuma-gui/pull/1907/commits/7752e4a7d8ee8234adcbe421392ce53f75405180 to trigger DataSource externally.

Do you think there is a better way?

Reply:
> Why does this change the behavior of DataSource?

It doesn't necessarily, the `sources` originally where a map of URIs to functions returning promises. That promise resolves the emitted `data` for the DataSource. Here the `async` turns the CallableEventSource created via `source()` into `return Promise.resolve(CallableEventSource)` instead of just `return CallableEventSource`. So the DataSource uses the CallableEventSource resolved by the `async`/Promise returned by the function as the `data` emitted by the DataSource.

Largely this is down to us supporting two methods for making `sources`:

1. Just return a promise which resolves your data. This approach is super simple and is enough so the majority of what we need to do is mostly a straight forwards one-liner, i.e. with little code.
2. You can also use `source(() => Promise<>, config)` which returns a CallableEventSource instead of a promise. This option lets you configure the CallableEventSource with more options instead of the defaults that are used if you just return a promise, i.e. polling, retries, backoff and probably lots of other things.

Hope that helps, feel free to ask if the above isn't clear.

---

> Do you think there is a better way?

I had a very quick look at this and the straight forward answer is "yes probably".

I think I added just normal polling when I did the retry stuff, but I'd have to look again, maybe I just added the scaffold for it, or built the retry stuff with polling in mind, I can't remember. All the below depends on whether the polling needs finishing off or not (I just took a quick look and it looks like you can just set `source(() => Promise, {interval: 1000})`)

Off the top of my head, the DataSource should just be set to poll eternally. Then in the view we add the DataSource to the page as usual and let it poll, check the `data.items` until the length hits `10` then remove the DataSource from the page (which will stop the polling).

I'm guessing that once the DataSource is no longer on the page we need to keep hold of its data? Which I'm guessing we can do by keeping the data using `onchange` and doing something in JS land.

---

Apart from the above approach, I've had the much spoken about DataLoader code here now for a while but not wanted to push it as yet. DataLoader maintains the `data` on "disconnection" (i.e an error after a successful load). So with that we could poll, then in the `sources` file check for `data.items.length === 10` and throw an error. The DataLoader will consider that a "disconnection", but we completely control with DataLoader what we show on "disconnection" so we don't have to show an error, and crucially we would still have access to the `data` and the polling would stop without removing the DataLoader from the page.

I would guess you might be able to do something like that ^ with DataSource also, as I think the polling would stop on error, but the `data` and `DataSource` could be kept on the page, maybe try that out and see.

---

Completely separately, how come we poll for all dataplanes instead of just ones in the `default` mesh?

Reply:
> Completely separately, how come we poll for all dataplanes instead of just ones in the default mesh?

We don’t know about the user’s meshes in this context and we also don’t care.

Reply:
Just wanted to note, while I appreciate you getting this migrated over, to DataSource and I'm glad we are doing it, we should probably leave this in draft for the moment and continue on with the rest of the data layer migration before revisiting this.

I can see that in order to use the data layer stuff in onboarding it needs to use DataSource (which is a good thing), but we can hold off on worrying about Onboarding until the rest of the migration is done.

Thanks for putting the draft up though!

Reply:
Just FYI, I completely changed the approach here to be somewhat close to what you suggested in an earlier comment.

I’m gonna resolve this thread now. If there’s anything related to discuss, let’s use a new one.

Reply:
Gotcha, I just took a very quick look and this is looking much better, thanks!